flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.parquet.utils;
+
+import org.apache.flink.formats.parquet.vector.position.CollectionPosition;
+import org.apache.flink.formats.parquet.vector.position.RowPosition;
+import org.apache.flink.formats.parquet.vector.type.Field;
+
+import shaded.parquet.it.unimi.dsi.fastutil.booleans.BooleanArrayList;
+import shaded.parquet.it.unimi.dsi.fastutil.longs.LongArrayList;
+
+import java.util.Optional;
+
+import static java.lang.String.format;
+
+/** Utils to calculate nested type position. */
+public class NestedPositionUtil {
+
+    /**
+     * Calculate row offsets according to column's max repetition level, definition level, value's
+     * repetition level and definition level. Each row has three situation:
+     * <li>Row is not defined,because it's optional parent fields is null, this is decided by its
+     *     parent's repetition level
+     * <li>Row is null
+     * <li>Row is defined and not empty.
+     *
+     * @param field field that contains the row column message include max repetition level and
+     *     definition level.
+     * @param fieldRepetitionLevels int array with each value's repetition level.
+     * @param fieldDefinitionLevels int array with each value's definition level.
+     * @return {@link RowPosition} contains collections row count and isNull array.
+     */
+    public static RowPosition calculateRowOffsets(
+            Field field, int[] fieldDefinitionLevels, int[] fieldRepetitionLevels) {
+        int rowDefinitionLevel = field.getDefinitionLevel();
+        int rowRepetitionLevel = field.getRepetitionLevel();
+        int nullValuesCount = 0;
+        BooleanArrayList nullRowFlags = new BooleanArrayList();
```

Comment:
You can add a `BooleanArrayList` just like  `org.apache.flink.runtime.util.LongArrayList`.